Sources/MongoSwift/SDAM.swift

```diff
@@ -231,29 +254,132 @@ public struct TopologyDescription {
         }
     }
 
+    /// An internal function used to calculate staleness of a given server in a replica set
+    internal func staleness(for server: ServerDescription) -> Double? {
+        guard [.replicaSetWithPrimary, .replicaSetNoPrimary].contains(self.type) else {
+            return nil
+        }
+
+        // From spec: Non-secondary servers have zero staleness
+        guard server.type == .rsSecondary else {
+            return 0.0
+        }
+
+        // Can force unwrap lastWriteDate in here since it's always present in the the ismaster output of primaries
+        // and secondaries, and we filter out all other types of servers.
+        if self.type == .replicaSetWithPrimary {
+            let primary = self.servers.filter { server in server.type == .rsPrimary }[0]
+
+            let delta = { (serverDesc: ServerDescription) -> Double? in
+                guard let serverLastUpdate = serverDesc.lastUpdateTime else {
+                    return nil
+                }
+                return (serverLastUpdate - serverDesc.lastWriteDate!) * 1000.0
+            }
+
+            guard let clientToServer = delta(server), let clientToPrimary = delta(primary) else {
+                return nil
+            }
+
+            return clientToServer - clientToPrimary + Double(TopologyDescription.heartbeatFrequencyMS)
+        } else { // ReplicaSetNoPrimary case
+            let secondaries = self.servers.filter { $0.type == .rsSecondary }
+            guard secondaries.count > 0 else {
+                return nil
+            }
+            let sMax = secondaries.max { $0.lastWriteDate! < $1.lastWriteDate! }!
+
+            return (sMax.lastWriteDate! - server.lastWriteDate!) + Double(TopologyDescription.heartbeatFrequencyMS)
```

Comment:
Oh, nice catch. I do convert it for the ReplicaSetPrimary case, but forgot to do so here. I think it makes more sense if we abide by Swift's decision to use seconds as the standard time unit instead of converting to MS everywhere anyways, so I've updated all `Double`s to be `TimeIntervals` and only converted `heartbeatFrequencyMS`. The value we're comparing this against is `maxStalenessSeconds` anyways.

This should've been caught by my testing and wasn't, so I'm going to add some of the spec tests in tomorrow to make it a little more robust.